provider/provider.go

```diff
@@ -464,37 +487,176 @@ func (s *SweepingProvider) provideKeysToPeer(p peer.ID, keys []mh.Multihash, pme
+	for _, prefixAndKeys := range sortedPrefixesAndKeys {
+		s.provideQueue.Enqueue(prefixAndKeys.Prefix, prefixAndKeys.Keys...)
+	}
+
+	go s.provideLoop()
+}
+
+// groupAndScheduleKeysByPrefix groups the supplied keys by their prefixes as
+// present in the schedule, and if `schedule` is set to true, add these
+// prefixes to the schedule to be reprovided.
+func (s *SweepingProvider) groupAndScheduleKeysByPrefix(keys []mh.Multihash, schedule bool) map[bitstr.Key][]mh.Multihash {
+	avgPrefixLen := -1
+	prefixes := make(map[bitstr.Key][]mh.Multihash)
+	seen := make(map[bit256.Key]struct{})
+
+	s.scheduleLk.Lock()
+	defer s.scheduleLk.Unlock()
+	for _, h := range keys {
+		k := keyspace.MhToBit256(h)
+		// Don't add duplicates
+		if _, ok := seen[k]; ok {
+			continue
+		}
+		seen[k] = struct{}{}
+
+		prefixConsolidation := false
+		prefix, scheduled := keyspace.FindPrefixOfKey(s.schedule, k)
+		if !scheduled {
+			if avgPrefixLen == -1 {
+				avgPrefixLen = s.getAvgPrefixLenNoLock()
+			}
+			prefix = bitstr.Key(key.BitString(k)[:avgPrefixLen])
+
+			if schedule {
+				if subtrie, ok := keyspace.FindSubtrie(s.schedule, prefix); ok {
+					// If generated prefix is a prefix of existing scheduled keyspace
+					// zones, consolidate these zones around the shorter prefix.
+					for _, entry := range keyspace.AllEntries(subtrie, s.order) {
+						s.schedule.Remove(entry.Key)
+					}
+					prefixConsolidation = true
+				}
+				s.schedulePrefixNoLock(prefix, false)
+			}
+		}
+
+		if _, ok := prefixes[prefix]; !ok {
+			prefixes[prefix] = []mh.Multihash{h}
+		} else {
+			prefixes[prefix] = append(prefixes[prefix], h)
+		}
+		if prefixConsolidation {
+			// prefix is a shorted prefix of a prefix already in the schedule.
+			// Consolidate everything into prefix.
+			for p, keys := range prefixes {
+				if keyspace.IsBitstrPrefix(p, prefix) {
+					seen := make(map[string]struct{})
+					for _, key := range prefixes[prefix] {
+						seen[string(key)] = struct{}{}
+					}
+					for _, key := range keys {
+						strKey := string(key)
+						if _, ok := seen[strKey]; ok {
+							continue
+						}
+						seen[strKey] = struct{}{}
+						prefixes[prefix] = append(prefixes[prefix], key)
+					}
+					delete(prefixes, p)
+				}
+			}
+		}
+	}
+	return prefixes
+}
+
+// provideLoop is the loop providing keys to the DHT swarm as long as the
+// provide queue isn't empty.
 //
-// Add the supplied multihashes to the provide queue, and return immediately.
-// The provide operation happens asynchronously.
-func (s *SweepingProvider) ProvideOnce(keys ...mh.Multihash) {
+// The s.provideRunning mutex prevents concurrent executions of the loop.
+func (s *SweepingProvider) provideLoop() {
+	if !s.provideRunning.TryLock() {
+		// Ensure that only one goroutine is running the provide loop at a time.
+		return
+	}
+	defer s.provideRunning.Unlock()
+
+	for !s.provideQueue.IsEmpty() {
+		if s.closed() {
+			// Exit loop if provider is closed.
+			return
+		}
```

Comment:
Instead of check `s.closed()` in the loop, consider making `close` wait for any `providerLoop` to exit and only checking `s.closed()` at the beginning of the function.